README.md
# clousre_lenin

A Clojure library designed to ... well, that part is up to you.

## Usage

FIXME

## License

Copyright © 2022 FIXME

This program and the accompanying materials are made available under the
terms of the Eclipse Public License 2.0 which is available at
http://www.eclipse.org/legal/epl-2.0.

This Source Code may also be made available under the following Secondary
Licenses when the conditions for such availability set forth in the Eclipse
Public License, v. 2.0 are satisfied: GNU General Public License as published by
the Free Software Foundation, either version 2 of the License, or (at your
option) any later version, with the GNU Classpath Exception which is available
at https://www.gnu.org/software/classpath/license.html.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2022-04-25
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2022-04-25
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/clousre_lenin/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/clousre_lenin/compare/0.1.0...0.1.1


test/clousre_lenin/core_test.clj
(ns clousre-lenin.core-test
  (:require [clojure.test :refer :all]
            [clousre-lenin.core :refer :all]))

(deftest a-test
  (testing "FIXME, I fail."
    (is (= 0 1))))


src/activities/actividad_5_1.clj
(ns activities.actividad-5-1)

; Rodrigo Alfredo Mendoza España A01720627
; Sergio Manuel González Vargas A01745446

(defn sign [n]
  (cond
    (= n 0) 0
    (< n 0) -1
    (> n 0) 1) )

(println(sign -5))
(println(sign 10))
(println(sign 0))


(defn roots [a b c]
  (/ (- (Math/sqrt (- (* b b) (* 4 a c))) b) (* 2 a)))

(println(roots 2 4 2))
(println(roots 1 0 0))
(println(roots 4 5 1))

(defn bmi [weight height]
  (def bmi-index (/ weight (* height height)))
  (cond
    (< bmi-index 20) "underweight"
    (and (< bmi-index 25) (> bmi-index 20)) "normal"
    (and (< bmi-index 30) (> bmi-index 25)) "obese1"
    (and (< bmi-index 40) (> bmi-index 30)) "obese2"
    (> bmi-index 40) "obese3"))

(println(bmi 45 1.7))
(println(bmi 55 1.5))
(println(bmi 76 1.7))
(println(bmi 81 1.6))
(println(bmi 120 1.6))

(defn factorial [x]
  (loop [n x resultado 1]
    (if (= n 0)
      resultado
      (recur (- n 1) (* n resultado)))))

(println(factorial 0))
(println(factorial 5))
(println(factorial 20))

(defn fib [n]
  (if (<= n -1)
    n
    (loop [contador n primero 0 segundo 1]
      (if (= contador 0)
        primero
        (recur (- contador 1) segundo (+ primero segundo))))))

(println(fib 6))
(println(map fib (range 10) ))
(println(fib 42))

(defn pow [a, b]
  (loop [numeroPotencia b resultado 1]
    (if (= numeroPotencia 0)
      resultado
      (recur (- numeroPotencia 1) (* resultado a)))))

(println(pow 5 0))
(println(pow -5 3))
(println(pow 15 12))

src/activities/actividad_5_2.clj
(ns activities.actividad-5-2)

; Rodrigo Alfredo Mendoza España A01720627
; Sergio Manuel González Vargas A01745446


;Actividad 5.2
;(println)
;1
;Con conj agrega el elemento n a la lista y sortea todos los elementos
(defn insert
  [n lst]
  (sort (conj lst n)))

(println "EJERCICIO 1")
(println(insert 14 '()))

(println(insert 4 '(5 6 7 8)))

(println(insert 5 '(1 3 6 7 9 16)))

(println(insert 10 '(1 5 6)))

;2
;Ingresa a listaNueva una lista vacia y a listaVieja el parametros
;Si listaVieja está vacia regresa listaNueva
;Utiliza recursión para ingresar a listaNueva los elementos de listaVieja y los sortea con insert
(defn insertion-sort
  [lst]
  (loop [listaNueva '() listaVieja lst]
    (if (empty? listaVieja)
      listaNueva
      (recur (insert (peek listaVieja) listaNueva) (pop listaVieja)))))

(println "EJERCICIO 2")
(println(insertion-sort '()))
(println(insertion-sort '(4 3 6 8 3 0 9 1 7) ))
(println(insertion-sort '(1 2 3 4 5 6)))
(println(insertion-sort '(5 5 5 1 5 5 5)))

;3
;Si n es mayor a 0, concatenar las dos listas creadas con split-at dependiendo del modulo de n con 7
(defn rotate-left
  [n lst]
  (if (= n 0)
    lst
    (if (> n 0)
      (let [[principio final] (split-at (mod n 7) lst)]
        (concat final principio)))))

(println "EJERCICIO 3")
(println(rotate-left 5 '()) )
(println(rotate-left 0 '(a b c d e f g)))
(println(rotate-left 1 '(a b c d e f g)))
(println(rotate-left 3 '(a b c d e f g)))
(println(rotate-left 8 '(a b c d e f g)))
(println(rotate-left 45 '(a b c d e f g)) )

;4
;Si b es igual a 0, regresa a
;Si b no es igual a 0, aplica la recursión de b con el remainer de a con b
(defn gcd
  [a b]
  (if (= b 0)
    a
    (recur b (rem a b))))

(println "EJERCICIO 4")
(println(gcd 13 7919) )
(println(gcd 20 16) )
(println(gcd 54 24))
(println(gcd 6307 1995))
(println(gcd 48 180))
(println(gcd 42 56))
;5
;Utilizando partition-by que hace que la lista se separe cada vez que hay un valor nuevo
;Utilizando map para agarrar todas las listas separadas, agarramos el primer valor
(defn compress
  [lst]
  (if (empty? lst)
    ()
    (map first (partition-by identity lst))))

(println "EJERCICIO 5")
(println(compress '()))
(println(compress '(a b c d)))
(println(compress '(a a a a b c c a a d e e e e)))
(println(compress '(a a a a a a a a a a)))
;6
;Crea una nueva función que toma los parametros 'x' y 'y' y regresa la funcion con los parametros en orden inverso
(defn args-swap
  [f]
  (fn [x y] (f y x)))

(println "EJERCICIO 6")
(println((args-swap list) 1 2) )
(println((args-swap /) 8 2) )
(println((args-swap cons) '(1 2 3) '(4 5 6)))
(println((args-swap map) '(-1 1 2 5 10) /))
;7
;Regresa la función de la derivada aplicando la función correspondiente
(defn deriv
  [f h]
  (fn [x]
    (/ (- (f (+ x h)) (f x)) h)))

(println "EJERCICIO 7")
(defn f [x] (* x x x))
(def df (deriv f 0.001))
(def ddf (deriv df 0.001))
(def dddf (deriv ddf 0.001))

(println(df 5))
(println(ddf 5))
(println(dddf 5))
;8
;Si n es cero, regresa 0
; Si n no es cero regresa método de newton aplicando la función correspondiente
(defn newton
  [f n]
  (if (= n 0)
    0
    (let [x-n-1 (newton f (dec n)) f' (deriv f 0.0001)]
      (- x-n-1 (/ (f x-n-1) (f' x-n-1))))))

(println "EJERCICIO 8")
(defn f1 [x] (- x 10))
(defn f2 [x] (+ (* 4 x) 2))
(defn f3 [x] (+ (* x x x) 1))
(defn f4 [x] (+(Math/cos x) (* 0.5 x)))

(println(newton f1 1) )
(println(newton f2 1))
(println(newton f3 50))
(println(newton f4 5))

src/clousre_lenin/core.clj
(ns clousre-lenin.core)

(defn foo
  "I don't do a whole lot."
  [x]
  (println x "Hello, World!"))

;Define una función average que recibe dos argumentos y regrese el promedio
(defn average
  [x y]
  (/ (+ x y)2))

;convertir de farenheight a celcius
(defn fahrenheit-to-celsius
  [f]
  (/ (* (- f 32) 5) 9))

;imprima mensaje con argumentos
(defn long-greeting
  [nombre edad carrera]
  (println "Nombre:" nombre "\nEdad:" edad "\nCarrera:" carrera))

;Actividad 5.1
(defn negative [x]
  (cond
    (= x 0) 0
    (< x 0) -1
    (> x 0) 1) )

(defn roots [a b c]
  (/ (- (Math/sqrt (- (* b b) (* 4 a c))) b) (* 2 a)))


(defn bmi [weight height]
  (def bmi-index (/ weight (* height height)))
  (cond
    (< bmi-index 20) "underweight"
    (and (< bmi-index 25) (> bmi-index 20)) "normal"
    (and (< bmi-index 30) (> bmi-index 25)) "obese1"
    (and (< bmi-index 40) (> bmi-index 30)) "obese2"
    (> bmi-index 40) "obese3"))

(defn fact [x]
  (loop [n x resultado 1]
    (if (= n 0)
      resultado
        (recur (- n 1) (* n resultado)))))

(defn fib [n]
  (if (<= n -1)
    n
      (loop [contador n primero 0 segundo 1]
        (if (= contador 0)
          primero
            (recur (- contador 1) segundo (+ primero segundo))))))


(defn pow [a, b]
  (loop [numeroPotencia b resultado 1]
    (if (= numeroPotencia 0)
      resultado
      (recur (- numeroPotencia 1) (* resultado a)))))

;Actividad 5.2

;1
;Con conj agrega el elemento n a la lista y sortea todos los elementos
(defn insert
  [n lst]
  (sort (conj lst n)))


;2
;Ingresa a listaNueva una lista vacia y a listaVieja el parametros
;Si listaVieja está vacia regresa listaNueva
;Utiliza recursión para ingresar a listaNueva los elementos de listaVieja y los sortea con insert
(defn insertion-sort
  [lst]
  (loop [listaNueva '() listaVieja lst]
    (if (empty? listaVieja)
      listaNueva
      (recur (insert (peek listaVieja) listaNueva) (pop listaVieja)))))


;3
;Si n es mayor a 0, concatenar las dos listas creadas con split-at dependiendo del modulo de n con 7
(defn rotate-left
  [n lst]
  (if (= n 0)
    lst
    (if (> n 0)
      (let [[principio final] (split-at (mod n 7) lst)]
        (concat final principio)))))


;4
;Si b es igual a 0, regresa a
;Si b no es igual a 0, aplica la recursión de b con el remainer de a con b
(defn gcd
  [a b]
  (if (= b 0)
    a
    (recur b (rem a b))))

;5
;Utilizando partition-by que hace que la lista se separe cada vez que hay un valor nuevo
;Utilizando map para agarrar todas las listas separadas, agarramos el primer valor
(defn compress
  [lst]
  (if (empty? lst)
    ()
    (map first (partition-by identity lst))))

;6
;Crea una nueva función que toma los parametros 'x' y 'y' y regresa la funcion con los parametros en orden inverso
(defn args-swap
  [f]
  (fn [x y] (f y x)))

;7
;Regresa la función de la derivada aplicando la función correspondiente
(defn deriv
  [f h]
  (fn [x]
    (/ (- (f (+ x h)) (f x)) h)))

;8
;Si n es cero, regresa 0
; Si n no es cero regresa método de newton aplicando la función correspondiente
(defn newton
  [f n]
  (if (= n 0)
    0
    (let [x-n-1 (newton f (dec n)) f' (deriv f 0.0001)]
      (- x-n-1 (/ (f x-n-1) (f' x-n-1))))))